extensions/amp-megaphone/0.1/amp-megaphone.js

```diff
@@ -0,0 +1,234 @@
+    );
+  }
+
+  /** @override */
+  buildCallback() {
+    this.updateBaseUrl_();
+  }
+
+  /** @override */
+  mutatedAttributesCallback(mutations) {
+    const playlist = mutations['data-playlist'];
+    const episode = mutations['data-playlist'];
+    if (playlist !== undefined || episode !== undefined) {
+      this.updateBaseUrl_();
+    }
+
+    const light = mutations['data-light'];
+    const sharing = mutations['data-sharing'];
+    const episodes = mutations['data-episodes'];
+    const start = mutations['data-start'];
+    const tile = mutations['data-tile'];
+    if (
+      light !== undefined ||
+      sharing !== undefined ||
+      episodes !== undefined ||
+      start !== undefined ||
+      tile !== undefined
+    ) {
+      if (this.iframe_) {
+        this.iframe_.src = this.getIframeSrc_();
+      }
+    }
+  }
+
+  /**@override*/
+  layoutCallback() {
+    const height = this.element.getAttribute('height');
+    const iframe = this.element.ownerDocument.createElement('iframe');
+
+    iframe.setAttribute('frameborder', 'no');
+    iframe.setAttribute('scrolling', 'no');
+
+    iframe.src = this.getIframeSrc_();
+
+    this.unlistenMessage_ = listen(
+      this.win,
+      'message',
+      this.handleMegaphoneMessages_.bind(this)
+    );
+
+    this.applyFillContent(iframe);
+    iframe.height = height;
+    this.element.appendChild(iframe);
+
+    this.iframe_ = iframe;
+
+    return this.loadPromise(iframe);
+  }
+
+  /** @override */
+  unlayoutCallback() {
+    if (this.iframe_) {
+      removeElement(this.iframe_);
+      this.iframe_ = null;
+    }
+    if (this.unlistenMessage_) {
+      this.unlistenMessage_();
+    }
+    return true; // Call layoutCallback again.
+  }
+
+  /**
+   * @return {string}
+   * @private
+   */
+  getIframeSrc_() {
+    this.updateBaseUrl_();
+    let url = this.baseUrl_ + '/';
+    const mediaid = userAssert(
+      this.element.getAttribute('data-playlist') ||
+        this.element.getAttribute('data-episode'),
+      'data-playlist or data-episode is required for <amp-megaphone> %s',
+      this.element
+    );
+
+    if (!this.isPlaylist_) {
+      url += mediaid + '/';
+    }
+
+    const hasLightTheme = this.element.getAttribute('data-light') === 'true';
+    const hasSharingDisabled =
+      this.element.getAttribute('data-sharing') === 'false';
+    const episodes = this.element.getAttribute('data-episodes');
+    const start = this.element.getAttribute('data-start');
+    const hasTile = this.element.getAttribute('data-tile') === 'true';
+
+    const queryParams = dict({
+      'p': this.isPlaylist_ ? mediaid : undefined,
+      'light': hasLightTheme || undefined,
+      'sharing': hasSharingDisabled ? 'false' : undefined,
+      'episodes': (this.isPlaylist_ && episodes) || undefined,
+      'start': (!this.isPlaylist_ && start) || undefined,
+      'tile': (!this.isPlaylist_ && hasTile) || undefined,
+    });
+
+    return addParamsToUrl(url, queryParams);
```

Comment:
Since URL is only used once, here, you can inline the value.

```
return addParamsToUrl( this.baseUrl_ + !this.isPlaylist_ ? '/mediaid/' : '/', queryParams);
```